align_system/utils/incontext_utils.py

```diff
@@ -0,0 +1,287 @@
+import json
+import torch
+import random
+from copy import deepcopy
+from abc import ABCMeta, abstractmethod
+
+from align_system.utils import adm_utils
+from align_system.utils import outlines_prompts_utils
+from align_system.utils.hydrate_state import hydrate_scenario_state
+from align_system.prompt_engineering.outlines_prompts import (
+    scenario_state_description_with_relevant_char_info,
+    comparative_kdma_score_prediction_prompt
+)
+
+
+class IncontextExampleGenerator(object, metaclass=ABCMeta):
+    '''
+    Abstract class for incontext example generator
+    Instances of this class have unique set_icl_datasets() functions for formatting prompt and reponses 
+    '''
+    @abstractmethod
+    def __init__(self,
+                 incontext_settings, 
+                 target_kdmas, 
+                 log,
+                 **kwargs):
+        self.incontext_settings = incontext_settings
+        self.target_kdmas = target_kdmas
+        self.log = log
+        self.set_icl_datasets()
+
+    @abstractmethod
+    def set_icl_datasets(self):
+        '''
+        Sets self.icl_datasets which contains all the ICL examples
+        This is specific to each instance of the class because the prompt and response will vary
+        read_icl_dataset_files() is a generic helper method for this step
+
+        The keys of self.icl_datasets are the 'kdma' keys from self.target_kdmas,
+        the values are a list of all ICL examples for that kdma,
+        each ICL example is a dictionary with keys: 'prompt', 'response'
+        For example: {kdma: [{'prompt':str, 'response':json}, ...], ...}
+        '''
+        self.icl_datasets = {}
+        pass
+
+    def read_icl_dataset_files(self):
+        '''
+        Helper function for set_icl_datasets() - reads dataset files and gets examples for target_kdmas
+        Returns incontext_data dictionary with format:
+            {kdma:[{state, actions, choices, kdma_values}, ...], ...}
+        '''
+        incontext_data = {}
+        # For each kdma
+        for target_kdma in self.target_kdmas:
+            sys_kdma_name = target_kdma['kdma']
+            # Add examples for each dataset file
+            dset_files = self.incontext_settings["datasets"][sys_kdma_name]
```

Comment:
It's possible `sys_kdma_name` does not exist in `self.incontext_settings["datasets"]`, we either need to A) check and raise an exception appropriately or alternatively, B) just iterate over the `self.incontext_settings["datasets"]` dictionary.

B could result in loading more data than necessary, but it also might mean we could load the dataset just once at the beginning instead of every time we select an action (no matter the target, as long as we have something for each target KDMA we're given)

Reply:
Good point, I went with option A here https://github.com/ITM-Kitware/align-system/pull/115/commits/48807cf3a519fba796786e9b845700dfae571dc6
When we initialize the generator we create the examples dictionary for all target KDMAs. So we are not loading more data than necessary but we also only load that dataset once, because then we can call `select_icl_examples()` with any of the target kdmas and it does not reload the dataset. Let me know if that makes sense or if I'm misunderstanding your point

Reply:
> only load that dataset once

Right now, we construct a `IncontextExampleGenerator` per `top_level_choose_action` call, so we load the dataset every time. We could potentially move the incontext settings to the ADM instance settings and construct the ICL generator (load the dataset) once in `__init__`. But I think that'd mean we want to go with option B then, since (in theory, but not in reality yet), our KDMA alignment target could be different each time we choose an action

Reply:
Moving the ICL generator creation to the ADM `__init__` makes sense in the context of the comparative regression ADM and would be better. But for the base outlines ADM, the ICL example responses are dependent on the target KDMA values (because we need to know which choice is correct) and we don't have access to the target KDMAs in `__init__`. Comparative regression on the other hand doesn't use the target value to create responses,  only the ground truth KDMA values. I could move it for comparative regression only, but also the incontext settings are currently defined in the `inference_kwargs`, so I'd have to move that as well (which wouldn't work for the base outlines). 

@eveenhuis and @dmjoy I thought I could fix this without a larger restructure but now I don't think so. I'm inclined to leave it as is now and maybe open an issue to think about this more down the line. Thoughts?

Reply:
Thanks for looking into it Jadie.  Yes I don't think it's worth more restructuring at this time and we can make a point to potentially revisit later on.  Unless @eveenhuis strongly objects, I'll go ahead and merge this in once the changelog has been update.

Reply:
I don't strongly object